doc/website/blog: terraform v0.5.0 announcement (#1598)
ariga/atlas: doc/website/blog/2023-04-20-terraform-v047.md

---
title: "Manage database schemas with Terraform in plain SQL"
authors: rotemtam
tags: [schema, migration, terraform]
---
:::info TL;DR

You can now use plain SQL to define the desired state of your database schema.
  
[See an example](#managing-database-schemas-in-plain-sql-using-terraform)

:::

> Earlier this week, we released [v0.4.7](https://github.com/ariga/terraform-provider-atlas/releases/tag/v0.4.7) 

Reviewer: Sounds like a big feature. Why not release a new minor version? v0.5.0

In previous versions of the Atlas Provider, the dev-database needed to be provided
by the user. This was a bit cumbersome, as the user needed to spin up a database (usually
by running a local Docker container), and then provide the connection string to it
in the `dev_url` field of the `atlas_schema` resource.

To improve this experience, we added support for the `docker://` driver, which allows
the user to only provide the database engine and version, and Atlas will spin up an
ephemeral container with the correct database engine and version. In addition, starting
this version, users may define the `dev_url` on the provider scope. For example:

```hcl
provider "atlas" {
  dev_url = "docker://mysql/8/myapp"
}
```

### Defining the desired state of the database schema in plain SQL

In earlier versions of the Atlas Provider, the `atlas_schema` resource required the user
to provide an [Atlas HCL](/atlas-schema/sql-resources) file which describes the desired
state of the database schema. Many users found this syntax, which resembles Terraform's
own, to be clean and concise. However, others disliked it and asked for a way to define

> the desired state in plain SQL.

Reviewer 2: Dislike but as well limited support on HCL. With SQL, we now support everything SQL does support out if the box, which atlas does not yet.

To support this use-case, and many others, we have announced 

> [support for "schema loaders"](http://localhost:3000/blog/2023/03/13/atlas-v010#schema-loaders) -

Reviewer 2: Remove localhost

components that can be used to load the desired schema from many kinds of sources
(such as plain SQL, an existing database, or the data-model of an ORM). To use this capability,
users may use the `atlas_schema` data source, which accepts a `url` field that points to
the desired schema. The scheme of this URL determines which schema loader will be used, for instance:

* `file://schema.sql` - loads the schema from a local SQL file.
* `mysql://root:pass@localhost:3306/myapp` - loads the schema from an existing MySQL database.
* `ent://service/ent/schema` - loads the schema from the schema of an [Ent](https://entgo.io/)
  project.

### Managing database schemas in plain SQL using Terraform

> In the following example, we will show to use Terraform and the Atlas provider to manage

Author: ```suggestion
In the following example, we will show how you can use Terraform and the Atlas provider to manage
```

a MySQL database schema in plain SQL.

> Let's start by creating a Terraform configuration file named `main.tf` installing two

Author: ```suggestion
Let's start by creating a Terraform file named `main.tf` installing the Atlas
```

> Terraform providers:

Author: ```suggestion
Terraform provider:
```

In a real-world scenario, you would probably use a managed database service such as AWS 
RDS or Google Cloud SQL, but for the purpose of brevity, a local database will suffice. Run:
```
docker run -d --name mysql -p 3306:3306 -e MYSQL_ROOT_PASSWORD=pass -e MYSQL_DATABASE=myapp mysql:8
```

Now that we have a database to manage, we can define the desired state of the database schema.
Add a file named "schema.sql" with the following content:

```sql
create table users (
    id   int          not null auto_increment primary key,
    name varchar(255) not null
);
```

Next, we will define an `atlas_schema` data source that will load the schema from the
`schema.sql` file:

```hcl
data "atlas_schema" "sql" {
  src = "file://${path.module}/schema.sql"
}
```

Finally, we will define an `atlas_schema` resource that will manage the schema in the
target database. In addition, we will configure the Atlas provider to use the `docker://`
driver to spin up a temporary database container:

```hcl
provider "atlas" {
  dev_url = "docker://mysql/8/myapp"
}

resource "atlas_schema" "mysql" {
  url = "mysql://root:pass@localhost:3306/myapp"

>   schema = data.atlas_schema.sql.schema

Reviewer 3: ```suggestion
  hcl = data.atlas_schema.sql.hcl
```

>   depends_on = [docker_container.mysql]

Reviewer 3: You don't have any resource for docker container in the example

}
```

> Now that we have defined our Terraform configuration, we can run `terraform init` to

Reviewer 3: Please add the full content of `main.tf` here to easy copy & paste

install the required providers:

```console
terraform init
```
This should output something like:
```
Initializing provider plugins...
- Finding ariga/atlas versions matching "0.4.7"...
- Installing ariga/atlas v0.4.7...
- Installed ariga/atlas v0.4.7 (signed by a HashiCorp partner, key ID 45441FCEAAC3770C)

# ...

Terraform has been successfully initialized!
```
Finally, we can run `terraform apply` to create the database schema:

```console
terraform apply
```

> Terrform will print the following plan:

Reviewer 2: ```suggestion
Terraform will print the following plan:
```

│   on main.tf line 18, in resource "atlas_schema" "myapp":
│   18: resource "atlas_schema" "myapp" {
│ 
│ The following SQL statements will be executed:
│ 
│ 
│ CREATE TABLE `myapp`.`groups` (`id` int NOT NULL AUTO_INCREMENT, `name` varchar(255) NOT NULL, PRIMARY KEY (`id`)) CHARSET
│ utf8mb4 COLLATE utf8mb4_0900_ai_ci
│ ALTER TABLE `myapp`.`users` ADD COLUMN `group_id` int NOT NULL, ADD INDEX `group_id` (`group_id`), ADD CONSTRAINT
│ `users_ibfk_1` FOREIGN KEY (`group_id`) REFERENCES `myapp`.`groups` (`id`) ON UPDATE NO ACTION ON DELETE NO ACTION
│ 
╵

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: 
```
After typing `yes` and pressing enter, Terraform will apply the changes, bringing
the actual state of the database schema in line with the desired state:

```
atlas_schema.myapp: Modifying... [id=mysql:///myapp]
atlas_schema.myapp: Modifications complete after 4s [id=mysql:///myapp]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

## Conclusion

In this tutorial, we have seen how to use Terraform to manage the schema of a MySQL database

> using the Atlas provider with plain SQL.  Using this workflow, teams can bridge the gap between

Reviewer 2: ```suggestion
using the Atlas provider with plain SQL. Using this workflow, teams can bridge the gap between
```
